# Pandas. Домашнее задание

В домашнем задании есть обязательные и дополнительные задачи.

*Обязательные задачи* нужно сделать, чтобы проверить, что вы действительно усвоили материал модуля. Они обязательны к загрузке на проверку преподавателю.

*Дополнительные задачи* для тех, кто чувствует, что нужно ещё немного потренироваться и порешать задачки по работе с Pandas. Вы также можете загрузить их на проверку преподавателю и получить по ним обратную связь.

Удачи!



---



---



**Цели домашнего задания:**
1. Научиться работать с описательными статистиками набора данных.
2. Научиться выбирать данные по условиям.
3. Научиться объединять несколько датафреймов в один.


**Что входит в домашнее задание:**
1. *Обязательные задачи*:

17 задач с использованием:
- набора рекламных данных, показывающих, щелкнул ли конкретный пользователь интернета рекламу на веб-сайте компании или нет,
- дополнительного набора данных, описывающего пользователей, посещающих веб-сайт компании.
2. *Дополнительные задачи*:

3 задачи для отработки исследования данных.


**Критерии оценки:**
- программа выдаёт верный ответ на заданном наборе данных;
- код читабелен: переменным даны осмысленные названия, отступы и правила расстановки пробелов соблюдены.

**Как отправить задание на проверку**

Выполните задачи, сохраните изменения при помощи опции Save and Checkpoint из вкладки меню File или кнопки Save and Checkpoint на панели инструментов. Отправьте через форму на платформе итоговый файл Jupyter Notebook (в формате .ipynb) или ссылку на него.



---


---



**Обязательные задачи**

В течение всего домашнего задания вы будете работать с несколькими наборами данных:
1. Набор рекламных данных, показывающих, щёлкнул ли конкретный пользователь рекламу на сайте компании или нет. Состоит из трёх разных файлов: `advertising_1.csv`, `advertising_2.csv`, `advertising_3.csv`.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Daily Time Spent on Site**: Среднее ежедневное время нахождения пользователя на сайте в минутах.
- **Daily Internet Usage**: Среднее ежедневное время нахождения пользователя в интернете в минутах.
- **Ad Topic Line**: Заголовок рекламного объявления.
- **Clicked on Ad**: Кликнул пользователь на рекламу или нет.

Дополнительный к рекламным данным датасет, который описывает пользователей, посещающих сайт компании. Состоит из одного файла: users.csv.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Age**: Возраст пользователя.
- **Male**: Является ли пользователь мужчиной.
- **City**: Город пользователя.
- **Country**: Страна пользователя.
- **Area Income**: Средний доход в месте жительства пользователя.


***Задача 1***

Загрузите таблицу из файла `advertising_1.csv` и сохраните её в датафрейм `adv1_df`. Укажите в качестве индекса столбец `Number`. Выведите на экран первые десять строк.


In [128]:
import pandas as pd
# Ваш код здесь
adv1_df = pd.read_csv("data/advertising_1.csv", index_col="Number")
adv1_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
525                        46.61                136.18   
526                        63.26                120.46   
527                        79.16                202.90   
528                        67.94                128.16   
529                        79.91                230.18   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
...                                       ...            ...  
525       Progressive intermediate throughput              1  
526             Customizable holistic archive              1  
527           Compatible intermediate concept              0  
528      Assimilated next generation firmware              1  
529        Total zero administration software              0  

[500 rows x 4 columns]

***Задача 2***

Выведите размер датафрейма `adv1_df`.

In [129]:
# Ваш код здесь
adv1_df.shape

(500, 4)

***Задача 3***

Выведите cреднее ежедневное время нахождения в интернете пользователя под номером 8.

In [130]:
# Ваш код здесь
# print(adv1_df.columns)
adv1_df.loc[8, 'Daily Internet Usage']


245.89

***Задача 4***

Загрузите таблицу из файла `advertising_2.csv` и сохраните её в датафрейм `adv2_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран строки для пользователей с номерами с 533-го по 536-й.

In [131]:
# Ваш код здесь
adv2_df = pd.read_csv("data/advertising_2.csv", index_col="Number")
adv2_df.loc[533:537]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
533                        46.61                   NaN   
534                          NaN                162.03   
535                        65.10                118.10   
536                        53.44                   NaN   
537                        79.53                244.91   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
533             Ergonomic methodical encoding              1  
534         Compatible dedicated productivity              0  
535            Up-sized real-time methodology              1  
536     Up-sized next generation architecture              1  
537           Managed 6thgeneration hierarchy              0

***Задача 5***

Выведите сводную таблицу описательных статистик для датафрейма `adv2_df`, вызвав необходимый метод на всём датафрейме. 

Глядя на вывод, напишите, столбец с каким названием не содержит незаполненных значений.

*Подсказка:*  В сводной таблице не содержится информации о количестве строк всего датафрейма, поэтому для сначала распечатайте общее количество строк датафрейма.

*Рекомендация:* Кроме изученного метода `describe` для описания датафрейма и получения информации о пропущенных значениях полезен метод [`info`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html).

In [132]:
# Ваш код здесь
adv2_df.describe()

Daily Time Spent on Site  Daily Internet Usage  Clicked on Ad
count                184.000000            184.000000     200.000000
mean                  66.494728            186.809674       0.420000
std                   15.414725             43.844580       0.494797
min                   32.910000            108.030000       0.000000
25%                   52.805000            148.955000       0.000000
50%                   71.150000            194.170000       0.000000
75%                   78.352500            225.372500       1.000000
max                   91.430000            259.760000       1.000000

In [133]:
# Ваш код здесь
print('Не содержит незаполненных значений столбец:', "Clicker on Ad")

Не содержит незаполненных значений столбец: Clicker on Ad


***Задача 6***

Добавьте к датафрейму `adv1_df` в конец датафрейм `adv2_df` с помощью метода `append`. 

Сохраните объединенный датафрейм в новой переменной `adv12_df` и распечатайте его.

In [134]:
# Ваш код здесь
adv12_df = adv1_df._append(adv2_df)
adv12_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
735                        65.53                190.17   
736                        61.87                   NaN   
737                        83.16                   NaN   
738                        44.11                121.24   
739                        56.57                131.98   

                                    Ad Topic Line  Clicked on Ad  
Number                                                            
5              Cloned 5thgeneration orchestration              0  
6              Monitored national standardization              0  
8           Triple-buffered reciprocal time-frame              0  
9                   Robust logistical utilization              0  
10                Sharable client-driven software              0  
...                                           ...            ...  
735        Pre-emptive client-server installation              0  
736     Multi-channeled attitude-oriented toolset              0  
737                 Decentralized 24hour approach              0  
738                Organic next generation matrix              1  
739          Multi-channeled non-volatile website              1  

[700 rows x 4 columns]

***Задача 7***

Для объединённого датафрейма `adv12_df` распечатайте среднее время, проводимое пользователем на сайте компании.

In [135]:
# Ваш код здесь
adv12_df["Daily Time Spent on Site"].std()

15.836567181601522

***Задача 8***

Из объединённого датафрейма `adv12_df` выберите и распечатайте строки только с теми пользователями, для которых неизвестно среднее ежедневное время, проводимое пользователем на сайте компании.

In [136]:
# Ваш код здесь
adv12_df.loc[adv12_df["Daily Internet Usage"].isna()]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
13                         74.53                   NaN   
21                         55.39                   NaN   
29                         41.39                   NaN   
46                         64.24                   NaN   
53                         44.33                   NaN   
...                          ...                   ...   
709                        76.02                   NaN   
719                        71.27                   NaN   
721                        39.96                   NaN   
736                        61.87                   NaN   
737                        83.16                   NaN   

                                    Ad Topic Line  Clicked on Ad  
Number                                                            
13                 Configurable coherent function              0  
21             Customizable multi-tasking website              1  
29                  Exclusive neutral parallelism              1  
46               Front-line intermediate database              0  
53        Visionary maximized process improvement              1  
...                                           ...            ...  
709         Distributed leadingedge orchestration              0  
719     Quality-focused bi-directional throughput              0  
721       Phased leadingedge budgetary management              1  
736     Multi-channeled attitude-oriented toolset              0  
737                 Decentralized 24hour approach              0  

[68 rows x 4 columns]

***Задача 9***

Загрузите таблицу из файла `advertising_3.csv` и сохраните её в датафрейм `adv3_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран столбцы `Ad Topic Line` и `Clicked on Ad`.

In [137]:
# Ваш код здесь
adv3_df = pd.read_csv("data/advertising_3.csv", index_col="Number")
adv3_df[["Ad Topic Line", "Clicked on Ad"]]

Ad Topic Line  Clicked on Ad
Number                                                             
740                 Distributed bifurcated challenge              0
741       Customizable zero-defect Internet solution              0
742     Self-enabling zero administration neural-net              1
744           Open-architected system-worthy ability              0
745               Quality-focused maximized extranet              1
...                                              ...            ...
999                    Front-line bifurcated ability              1
1000                   Fundamental modular algorithm              1
1001                 Grass-roots cohesive monitoring              1
1002                    Expanded intangible solution              1
1003            Proactive bandwidth-monitored policy              0

[251 rows x 2 columns]

***Задача 10***

Добавьте к датафрейму `adv12_df` в конец датафрейм `adv3_df` с помощью метода `concat`. Явно пропишите значение параметра `axis`.

Сохраните объединённый датафрейм в новой переменной `adv123_df` и распечатайте его.

In [138]:
# Ваш код здесь
adv123_df = pd.concat([adv12_df, adv3_df], axis=0)
adv123_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   
1003                       55.55                187.95   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
...                                       ...            ...  
999             Front-line bifurcated ability              1  
1000            Fundamental modular algorithm              1  
1001          Grass-roots cohesive monitoring              1  
1002             Expanded intangible solution              1  
1003     Proactive bandwidth-monitored policy              0  

[951 rows x 4 columns]

***Задача 11***

Для объединённого датафрейма `adv123_df` выберите только те строки, в которых пользователи кликнули на рекламное сообщение. Сохраните полученный результат в датафрейм `success_adv_df`, распечатайте размер нового датафрейма.

In [139]:
# Ваш код здесь
succes_adv_df = adv123_df[adv123_df["Clicked on Ad"] != 0]
succes_adv_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
12                           NaN                131.76   
15                         47.64                122.02   
17                         69.57                113.12   
19                           NaN                143.56   
20                         63.45                140.64   
...                          ...                   ...   
997                        69.17                123.62   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
12                   Reactive local challenge              1  
15             Centralized neutral neural-net              1  
17      Centralized content-based focus group              1  
19              Grass-roots coherent extranet              1  
20         Persistent demand-driven interface              1  
...                                       ...            ...  
997             Extended leadingedge solution              1  
999             Front-line bifurcated ability              1  
1000            Fundamental modular algorithm              1  
1001          Grass-roots cohesive monitoring              1  
1002             Expanded intangible solution              1  

[475 rows x 4 columns]

***Задача 12***

Загрузите таблицу с данными пользователей из файла `users.csv` и сохраните её в датафрейм `users_df`. Индекс не указывайте.

Распечатайте сам датафрейм и сводную таблицу со статистиками для него.

In [140]:
# Ваш код здесь
users_df = pd.read_csv("data/users.csv")
print(users_df)
users_df.describe()

     Number   Age  Male            City                 Country  Area Income
0         5   NaN   0.0     Wrightburgh                 Tunisia     61833.90
1         6  31.0   NaN       West Jodi                   Nauru     68441.85
2         7  26.0   0.0        Davidton              San Marino     59785.94
3         8  29.0   1.0  West Terrifurt                   Italy     54806.18
4         9   NaN   0.0    South Manuel                 Iceland     73889.99
..      ...   ...   ...             ...                     ...          ...
975    1000  30.0   NaN       Duffystad                 Lebanon     71384.57
976    1001  45.0   1.0     New Darlene  Bosnia and Herzegovina     67782.17
977    1002  51.0   NaN   South Jessica                Mongolia     42415.72
978    1003  19.0   0.0     West Steven               Guatemala     41920.79
979    1004  26.0   0.0     Ronniemouth                  Brazil     29875.80

[980 rows x 6 columns]


Number         Age        Male   Area Income
count   980.000000  860.000000  868.000000    903.000000
mean    504.485714   36.073256    0.500000  55010.631938
std     287.956480    8.795992    0.500288  13423.710892
min       5.000000   19.000000    0.000000  13996.500000
25%     255.750000   29.000000    0.000000  46947.570000
50%     503.500000   35.000000    0.500000  57032.360000
75%     753.250000   42.000000    1.000000  65537.990000
max    1004.000000   61.000000    1.000000  79484.800000

***Задача 13***

Глядя на статистики `users_df`, укажите значения, удовлетворяющие следующим условиям:
1. Возраст самого молодого пользователя сайта: *ваш_ответ*
2. Максимальный средний доход в месте жительства пользователя сайта: *ваш_ответ*
3. Средний возраст пользователя сайта: *ваш_ответ*
4. Количество заполненных значений для колонки, содержащей информацию о том, является пользователь мужчиной или нет: *ваш_ответ*
5. Минимальный номер пользователя: *ваш_ответ*

В этом задании не требуется выбирать по условиям и распечатывать значения, достаточно внимательно посмотреть на статистики и скопировать нужные значения согласно вопросам.

In [141]:

print("Min age:", 19)
print("Max income:", 79484.8)
print("Avg age", 8.795992)
print("Count Male isNotNa:", 868)
print("Min number:", 5)

# print("Min age:", users_df["Age"].min())
# print("Max income:", users_df["Area Income"].max())
# print("Avg age", users_df["Age"].std())
# print("Count Male isNotNa:", users_df["Male"].count())
# print("Min number:", min(users_df.index))


Min age: 19
Max income: 79484.8
Avg age 8.795992
Count Male isNotNa: 868
Min number: 5


***Задача 14***

К полученному датафрейму с успешными рекламными сообщениями `success_adv_df` добавьте справа информацию из датафрейма с пользователями `users_df` и сохраните в датафрейм `success_full_df`. В полученном датафрейме должны остаться только те пользователи, номера которых есть в обоих датафреймах. Для этого воспользуйтесь методом `merge` и пропишите явно верные параметры.

Затем распечатайте датафрейм.

*Важно:*

Объединять датафреймы с помощью `merge` можно и по индексам, и по столбцам с разными названиями. 
Так как общие значения содержатся в столбце `Number`, но в датафрейме `users_df` данный столбец является обычным столбцом, а в `success_adv_df` является индексом, перед тем как объединять, переведите столбец `Number` в `success_adv_df` в обычный столбец.

Для этого перед объединением пропишите:
```
success_adv_df = success_adv_df.reset_index()
```

[Подробнее о reset_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)

In [142]:
# Ваш код здесь
# succes_adv_df = succes_adv_df.reset_index()
succes_full_df = pd.merge(left=succes_adv_df, right=users_df, on='Number', how="inner")
succes_full_df
#users_df

Number  Daily Time Spent on Site  Daily Internet Usage  \
0        12                       NaN                131.76   
1        15                     47.64                122.02   
2        17                     69.57                113.12   
3        19                       NaN                143.56   
4        20                     63.45                140.64   
..      ...                       ...                   ...   
459     997                     69.17                123.62   
460     999                     43.70                173.01   
461    1000                     72.97                208.58   
462    1001                     51.30                   NaN   
463    1002                     51.63                120.37   

                             Ad Topic Line  Clicked on Ad   Age  Male  \
0                 Reactive local challenge              1  48.0   1.0   
1           Centralized neutral neural-net              1  49.0   NaN   
2    Centralized content-based focus group              1  48.0   1.0   
3            Grass-roots coherent extranet              1  33.0   0.0   
4       Persistent demand-driven interface              1  23.0   NaN   
..                                     ...            ...   ...   ...   
459          Extended leadingedge solution              1  40.0   NaN   
460          Front-line bifurcated ability              1  28.0   0.0   
461          Fundamental modular algorithm              1  30.0   NaN   
462        Grass-roots cohesive monitoring              1  45.0   1.0   
463           Expanded intangible solution              1  51.0   NaN   

                 City                 Country  Area Income  
0    Port Jefferybury               Australia     24593.33  
1     West Brandonton                   Qatar     45632.51  
2      West Katiefurt                   Egypt     51636.92  
3                 NaN                Barbados     30976.00  
4                 NaN                   Spain     52182.23  
..                ...                     ...          ...  
459       New Patrick              Montenegro     66193.81  
460      Nicholasland                 Mayotte     63126.96  
461         Duffystad                 Lebanon     71384.57  
462       New Darlene  Bosnia and Herzegovina     67782.17  
463     South Jessica                Mongolia     42415.72  

[464 rows x 10 columns]

***Задача 15***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, отобразите сводную таблицу со статистиками для строковых колонок (`Ad Topic Line`, `City`, `Country`).


In [143]:
# Ваш код здесь
succes_full_df[["Ad Topic Line", "City" ,"Country"]].describe()

Ad Topic Line        City   Country
count                        464         420       418
unique                       464         412       203
top     Reactive local challenge  Lake James  Ethiopia
freq                           1           2         7

***Задача 16***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите страну и количество вхождений этой страны в датафрейм.


In [144]:
# Ваш код здесь
#succes_full_df.groupby(["Country"])["Country"].agg(['count'])
succes_full_df['Country'].value_counts()
#succes_full_df['Country'].value_counts().nlargest(1).index[0]

Country
Ethiopia          7
Australia         6
Turkey            6
Liberia           6
Senegal           5
                 ..
Ukraine           1
United Kingdom    1
Montserrat        1
Timor-Leste       1
Comoros           1
Name: count, Length: 203, dtype: int64

***Задача 17***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных.


In [145]:
# Ваш код здесь
_max_country = succes_full_df['Country'].value_counts().nlargest(1).index[0]
succes_full_df[succes_full_df["Country"] == _max_country]

Number  Daily Time Spent on Site  Daily Internet Usage  \
245     525                     46.61                136.18   
246     526                     63.26                120.46   
278     606                     52.13                118.27   
297     650                       NaN                137.97   
299     652                     46.43                137.20   
417     919                     41.86                128.62   
423     930                       NaN                   NaN   

                                       Ad Topic Line  Clicked on Ad   Age  \
245              Progressive intermediate throughput              1  42.0   
246                    Customizable holistic archive              1  29.0   
278                 Open-source even-keeled database              1  50.0   
297      Seamless bandwidth-monitored knowledge user              1  28.0   
299  Reverse-engineered background Graphic Interface              1  28.0   
417                      Open-source stable paradigm              1  39.0   
423           Front-line fresh-thinking installation              1  45.0   

     Male               City   Country  Area Income  
245   0.0      Catherinefort  Ethiopia     65856.74  
246   1.0         East Donna  Ethiopia     54787.37  
278   1.0  West Raymondmouth  Ethiopia     40926.93  
297   1.0          Markhaven  Ethiopia     39723.97  
299   1.0          Brianland  Ethiopia     53922.43  
417   1.0        Hawkinsbury  Ethiopia     53041.77  
423   0.0     Huffmanchester  Ethiopia     43241.88

***Задача 18***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных, они проводят в интернете в среднем больше 120 минут и их возраст меньше 30 лет. 
Cохраните получившиеся строки в новый датафрейм `filtered_df`.


In [146]:
# Ваш код здесь
filtered_df = succes_full_df[((succes_full_df["Daily Internet Usage"] > 120) & (succes_full_df["Age"] < 30))]
filtered_df

Number  Daily Time Spent on Site  Daily Internet Usage  \
4        20                     63.45                140.64   
12       38                     55.60                212.58   
75      162                     41.73                202.18   
95      210                     58.22                120.90   
100     220                       NaN                151.95   
112     238                     78.32                239.52   
121     254                     62.20                161.16   
131     271                     70.58                136.94   
132     273                     40.17                171.31   
134     280                     47.66                156.54   
141     294                     60.53                167.22   
148     310                     79.81                178.85   
150     317                     75.55                169.40   
160     349                     61.09                131.68   
162     354                     60.91                184.94   
175     389                     49.67                153.69   
180     402                     50.52                219.69   
185     412                     48.73                142.04   
198     433                     60.53                123.28   
208     455                     43.41                160.73   
209     456                     55.92                145.08   
223     484                     39.76                196.83   
246     526                     63.26                120.46   
262     578                     74.59                158.35   
269     589                     56.16                164.25   
271     594                     39.86                161.24   
272     595                       NaN                141.13   
289     630                     67.91                146.80   
297     650                       NaN                137.97   
299     652                     46.43                137.20   
302     666                     46.28                228.78   
309     681                     39.36                161.79   
311     686                     44.15                141.96   
318     711                     71.33                169.40   
329     739                     56.57                131.98   
335     751                     66.80                248.51   
341     763                     36.56                195.89   
344     768                     52.35                147.61   
361     805                     56.64                123.24   
369     815                     43.16                143.04   
371     822                     54.47                141.52   
392     860                     50.87                190.41   
397     880                     40.34                173.75   
405     896                     54.39                171.90   
437     956                     41.35                162.46   
439     958                     45.53                141.58   
442     965                     56.78                124.32   
447     975                     40.18                151.96   
460     999                     43.70                173.01   

                                        Ad Topic Line  Clicked on Ad   Age  \
4                  Persistent demand-driven interface              1  23.0   
12         Multi-layered 4thgeneration knowledge user              1  23.0   
75               Sharable dedicated Graphic Interface              1  28.0   
95        User-friendly client-server instruction set              1  29.0   
100                      Front-line tangible alliance              1  23.0   
112                         Synergized cohesive array              1  28.0   
121               Cross-platform multimedia algorithm              1  25.0   
131          Proactive next generation knowledge user              1  26.0   
132                   Visionary asymmetric encryption              1  26.0   
134                     Devolved responsive structure              1  29.0

***Задача 19***

Для получившегося датафрейма `filtered_df` выведите строки с данными пользователей, для которых либо не указано среднее время, проведённое на сайте, либо указанное время больше 55 минут. Cохраните получившиеся строки в новый датафрейм `filtered_df2`.


In [147]:
# Ваш код здесь
filtered_df[((filtered_df["Daily Time Spent on Site"].isna() == False) | (filtered_df["Daily Time Spent on Site"] > 55))]

Number  Daily Time Spent on Site  Daily Internet Usage  \
4        20                     63.45                140.64   
12       38                     55.60                212.58   
75      162                     41.73                202.18   
95      210                     58.22                120.90   
112     238                     78.32                239.52   
121     254                     62.20                161.16   
131     271                     70.58                136.94   
132     273                     40.17                171.31   
134     280                     47.66                156.54   
141     294                     60.53                167.22   
148     310                     79.81                178.85   
150     317                     75.55                169.40   
160     349                     61.09                131.68   
162     354                     60.91                184.94   
175     389                     49.67                153.69   
180     402                     50.52                219.69   
185     412                     48.73                142.04   
198     433                     60.53                123.28   
208     455                     43.41                160.73   
209     456                     55.92                145.08   
223     484                     39.76                196.83   
246     526                     63.26                120.46   
262     578                     74.59                158.35   
269     589                     56.16                164.25   
271     594                     39.86                161.24   
289     630                     67.91                146.80   
299     652                     46.43                137.20   
302     666                     46.28                228.78   
309     681                     39.36                161.79   
311     686                     44.15                141.96   
318     711                     71.33                169.40   
329     739                     56.57                131.98   
335     751                     66.80                248.51   
341     763                     36.56                195.89   
344     768                     52.35                147.61   
361     805                     56.64                123.24   
369     815                     43.16                143.04   
371     822                     54.47                141.52   
392     860                     50.87                190.41   
397     880                     40.34                173.75   
405     896                     54.39                171.90   
437     956                     41.35                162.46   
439     958                     45.53                141.58   
442     965                     56.78                124.32   
447     975                     40.18                151.96   
460     999                     43.70                173.01   

                                        Ad Topic Line  Clicked on Ad   Age  \
4                  Persistent demand-driven interface              1  23.0   
12         Multi-layered 4thgeneration knowledge user              1  23.0   
75               Sharable dedicated Graphic Interface              1  28.0   
95        User-friendly client-server instruction set              1  29.0   
112                         Synergized cohesive array              1  28.0   
121               Cross-platform multimedia algorithm              1  25.0   
131          Proactive next generation knowledge user              1  26.0   
132                   Visionary asymmetric encryption              1  26.0   
134                     Devolved responsive structure              1  29.0   
141                       Organic interactive support              1  24.0   
148               Proactive context-sensitive project              1  24.0   
150                  Upgradable even-keeled challenge              1  22.0   
160             User-friendly

**Дополнительные задачи**

***Задача 20***

Для датафрейма `adv123_df` выведите строки с данными, заголовок рекламного сообщения в которых включает слово `model`.

Для этого вам понадобится метод [`contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).

Cохраните получившиеся строки в новый датафрейм `model_df`.


In [159]:
# Ваш код здесь
model_df = adv123_df[adv123_df["Ad Topic Line"].str.contains('model', regex=True)]
model_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
62                         44.98                129.31   
119                        69.00                   NaN   
202                        78.57                   NaN   
211                        76.90                212.67   
314                          NaN                161.16   
350                        65.77                   NaN   
501                        81.10                216.49   
538                        91.43                209.91   
629                        78.17                192.27   
712                        81.90                225.47   
823                        81.98                212.88   
995                        35.79                165.62   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
62                   Organized global model              1  
119                 Virtual composite model              0  
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
314              Extended interactive model              1  
350     Function-based fault-tolerant model              0  
501                Front-line dynamic model              0  
538                Organic motivating model              0  
629                  Automated mobile model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0  
995          Enterprise-wide tangible model              1

***Задача 21***

Для получившегося датафрейма `model_df` выведите строки с данными пользователей, которые проводят на сайте больше среднего времени по всем пользователям в данном датафрейме. Cохраните получившиеся строки в новый датафрейм `model_popular_df`.


In [178]:
# Ваш код здесь
print(model_df["Daily Time Spent on Site"].mean())
model_popular_df = model_df[model_df["Daily Time Spent on Site"] > model_df["Daily Time Spent on Site"].mean()]
model_popular_df

71.41727272727272


Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
202                        78.57                   NaN   
211                        76.90                212.67   
501                        81.10                216.49   
538                        91.43                209.91   
629                        78.17                192.27   
712                        81.90                225.47   
823                        81.98                212.88   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
202           Exclusive client-driven model              0  
211           Synchronized multimedia model              0  
501                Front-line dynamic model              0  
538                Organic motivating model              0  
629                  Automated mobile model              0  
712     Multi-channeled 3rdgeneration model              0  
823                Up-sized maximized model              0

***Задача 22***

Для объединённого датафрейма `model_popular_df` выведите строки с данными пользователей, номера которых равны 501, 712 и 823. 

Для этого вам понадобится метод [`isin`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html).

*Подсказка:* Не забудьте, что номера пользователей находятся не в обычной колонке, а в колонке, являющейся индексом, поэтому вам необходимо получить значения индекса, а не обычной колонки.

In [184]:
# Ваш код здесь
print(model_popular_df.index)
model_popular_df.isin(range(501))

Index([202, 211, 501, 538, 629, 712, 823], dtype='int64', name='Number')


Daily Time Spent on Site  Daily Internet Usage  Ad Topic Line  \
Number                                                                  
202                        False                 False          False   
211                        False                 False          False   
501                        False                 False          False   
538                        False                 False          False   
629                        False                 False          False   
712                        False                 False          False   
823                        False                 False          False   

        Clicked on Ad  
Number                 
202              True  
211              True  
501              True  
538              True  
629              True  
712              True  
823              True